### PR TITLE
make mv recover times and recover_policy configurable

### LINKED ISSUE #672: add a config setting for best effort recover times.
**Describe what enhancement you'd like to have**

### PR DESCRIPTION
**Pull request by yokofly**
Please write user-readable short description of the changes:
fix #672

### CODE REVIEW

File: src/Storages/Streaming/StorageMaterializedView.cpp [hunk 1]
```diff
@@ -537,7 +537,7 @@ void StorageMaterializedView::initBackgroundState()
                                 auto current_failed_sns = getProcessedSNOfStreamingSources(streaming_sources);
                                 recover_sns.clear();
                                 /// If the current SN fails three times in a row, we will skipping some SNs and recover from the next SN to avoid permanent recovery.
-                                if (failed_sn_queue.size() >= 2)
+                                if (failed_sn_queue.size() >= local_context->getSettingsRef().recovery_retry_for_sn_failure)
```
Comment (yl-lisen) [inline src/Storages/Streaming/StorageMaterializedView.cpp]: `local_context->getSettingsRef().recovery_retry_for_sn_failure - 1`

Comment (yokofly) [reply src/Storages/Streaming/StorageMaterializedView.cpp]: sorry I forget

Comment (yl-lisen) [reply src/Storages/Streaming/StorageMaterializedView.cpp]: In case `recovery_retry_for_sn_failure == 0`, need special handling

Comment (yokofly) [reply src/Storages/Streaming/StorageMaterializedView.cpp]: thanks, I submit new commit, plz check

File: src/Storages/Streaming/StorageMaterializedView.cpp [hunk 2]
```diff
@@ -529,15 +534,21 @@ void StorageMaterializedView::initBackgroundState()
...
                             {
                                 auto current_failed_sns = getProcessedSNOfStreamingSources(streaming_sources);
                                 recover_sns.clear();
-                                /// If the current SN fails three times in a row, we will skipping some SNs and recover from the next SN to avoid permanent recovery.
-                                if (failed_sn_queue.size() >= 2)
+
+                                /// If the current SN fails three times(can be adjusted in config) in a row, we will skipping some SNs and recover from the next SN to avoid permanent recovery.
+                                size_t recovery_retry_from_settings = local_context->getSettingsRef().recovery_retry_for_sn_failure;
+                                size_t recover_retry_times = DEFAULT_RECOVER_RETRY_FOR_SN_FAILURE;
+                                if (recovery_retry_from_settings > 0)
```
Comment (yokofly) [inline src/Storages/Streaming/StorageMaterializedView.cpp:548]: current code if the user sets the config as zero, will skip this branch, and then use the DEFAULT(3).
so you mean, the user set the recover policy as best_effort and then set the recovery_retry_for_sn_failure=0.
what does this user want?

File: src/Storages/Streaming/StorageMaterializedView.cpp [hunk 3]
```diff
@@ -529,15 +534,21 @@ void StorageMaterializedView::initBackgroundState()
...
+                                    recover_retry_times = recovery_retry_from_settings - 1;
+
+                                if (failed_sn_queue.size() >= recover_retry_times)
```
Comment (yl-lisen) [inline src/Storages/Streaming/StorageMaterializedView.cpp]: suggested changes
```c++
auto recovery_retry_for_sn_failure = 3;
...
if (failed_sn_queue.size() >= recover_retry_times - 1)
...
```

**Review by yl-lisen — APPROVED** [review]

**Pull request merged**